### ISSUE TITLE
Sequence.remove() not working properly

### ISSUE THREAD

**Issue by MikeSult** [Tonejs/Tone.js#269]
In regards to Pat Hine's question regarding Sequence.remove()

I think I found a subtle bug that is preventing seq.remove() from removing a subarray at the top level.  (Currently you need to use seq.at(1).remove(0)... as seq.remove(1) doesn't do anything on subarrays)  Here are my findings and hopefully I'm understanding the code correctly. 

When removing a subarray there is a recursive action involving both Part.remove() and Sequence.remove().  
1) In Part.remove() method It seems there is a type change in the type of 'time' from type Tone.TransportTime into type Number
line 17132
time = this.toTicks(time); 

2) later when Part.remove() encounters an array it calls Sequence.remove()

3) Sequence.remove() is called on a subarray [type Sequence])
approx line 17610
Tone.Sequence.prototype.remove(index, value) is called which calls 

4) Tone.Part.prototype.remove(this. this._indexTime(index), value) 
approx line:17126

5) and in Tone.Part.prototype.remove() the loop thru this._events[] looks for (event instanceof Tone.Part) if found it calls

event.remove(time, value)  

I thought it would call itself, Part.remove() (thinking this event was a Part) But console.log(event) was telling me that although it is an instanceof Tone.Part it's actually subclass of Part i.e. Sequence.

6) so it calls Tone.Sequence.prototype.remove(index, value) AGAIN but it is passed in time for the index param. But that should have parameters of (index, value) not (time, value)  

7) Then calls Tone.Part.prototype.remove(this. this._indexTime(index), value) AGAIN.  

So I traced the (index->time) value as this recursion takes place, and the time value gets corrupted by the calculation of this._indexTime(time) instead of this._indexTime(index).

I see that if you pass in time to ._indexTime(index) it looks to see the type of the index param (index instanceof Tone.TransportTime) but during the initial call to Tone.Part...  time = this.toTicks(time) is called which turns time into type Number instead of type Tone.TransportTime so now it doesn't recognize that the time param is 'time' and NOT an index.  It treats time as an index and time gets miscalculated and matches are never found.

With some experiments I found that even if the time value is correct, it still removed only the first element of the subarray so to delete the other elements, another loop is needed (using the j index):

Tone.js line 17132 ...
            time = this.toTicks(time);
	        for (var i = this._events.length - 1; i >= 0; i--) {
	            var event = this._events[i];
	            if (event instanceof Tone.Part) {
//------------------------ new code -------------------
	                for(var j = event.length - 1; j >= 0; j--) {
	                    event.remove(j, value);
	                }
// ------------------------- end new code ---------------------
//	                event.remove(time, value);

	            } else {
                 '''
                }

And it appears now .remove() will work with any size subarray IF it doesn't have nested arrays inside.

It DOES NOT work if the subarray has a sub-subarray.  I'm still trying to work that out.  A subArrayOffset is needed to be included in the if(event.startOffset === time) test inside the 'else' section to properly calibrate it so matches are recognized.

For A/B comparison, I've made a codepen (like Pat Hime's) showing the current behavior (which doesn't remove sequences).

https://codepen.io/mikesult/pen/JOGZRZ

The demo will remove sequences.  Unfortunately I couldn't make a codepen of the demo version because codpen doesn't allow links to script using http (only https) and my modified Tone.js version is on a http server.  So here's a link to the demo.

http://www.guitarland.com/ExperimentWithSeq.html

Both pages are reading the same javascript.

Mike Sult

P.S. Tone.js is bringing me great joy.
----------

Bugs and feature requests only please.

For help questions, check out the forum: https://groups.google.com/forum/#!forum/tonejs

If possible, please include a link which illustrates the issue.

https://jsfiddle.net/3m9k17ef/ (latest release)
https://jsfiddle.net/kyqnbr7d/ (dev build)

**Comment by MikeSult** [comment]
I think I wasn't complete in my suggested change in Tone.Part.remove() 
It still needs to have a branch between sequence and parts incase of nested parts (is that a thing?)  I've tested this with basic Parts and Sequences. and it hasn't changed or broken the Parts behavior and it now deletes sequences with a top level remove().

Tone.js line 17132 ...
	time = this.toTicks(time);
	for (var i = this._events.length - 1; i >= 0; i--) {
		var event = this._events[i];
		if (event instanceof Tone.Part) {
// ------------------------- new code -------------------------
			if(event instanceof Tone.Sequence) {
				for(var j = event.length -1; j >= 0; j--) {
					event.remove(j, value);
				}
			} else { // Tone.Part
				event.remove(time, value);
			}
// ------------------------- end new code -------------------------
//			event.remove(time, value);
		} else {
		 '''
		}

Still don't have a suggestion for nested sequences being completely removed with a top level remove()  It seems that the definition of time=0  in the inner-inner sequence is based on it's parent's startOffset not the main outer sequence's definition of time=0;  I've been experimenting with using a subArrayOffset when I recognize a subarray but I don't have it working properly yet.

cheers,
Mike Sult

**Comment by MikeSult** [comment]
Update:  Making some progress on nested sequences (sub-sub-arrays but no deeper so far)

Current solution isn't ready for primetime as I added a global
var subArrayOffset = 0;

just outside of the Tone.Part.prototype.remove()
which is used inside of Tone.Part.remove() when sub-sub-array are found.
Looking for a better way, discussion below.

Also found another issue regarding nulls in a simple sequence.  Used for rests they work fine but when removing them there is a problem.

CAVEAT: if there is a null in the subarray the last note isn't removed.  If there are more than one nulls in the sequence then that number of notes after the last null won't be removed. 

REASON: the sub-array.length attribute doesn't count the nulls so the loop using sub-array.length as the loop limit doesn't remove all of the sub-array. This seems to different behavior then regular javascript arrays:

var a = [null, "C3", "C3"]; 
a.length; // returns 3

However:
var seq = new Tone.Sequence(function(time, note){
	  synth.triggerAttackRelease(note, '8n', time);
	 }, [null, "C3", "C3"]).start();
	seq.length; // returns 2

So to completely remove sequences that have nulls there needs to be someway to count how many nulls are there.  I'm not sure how to do that in this situation.

remove() for sub-sub-arrays
Current behavior removes sub-sub-arrays (nulls still a problem here too).  
ANOTHER CAVEAT: if the are 3 or more sub-sub-arrays in the sub-array the last sub-sub-array must be at least 3 elements long (I have no idea why);

I'm trying to update the startOffset of the subArray elements but I can't seem to identify them uniquely.
DIFFICULTY: for some reason sub-sub-arrays are not recognized as (instance of Tone.Sequence) and instead show as null (??).  So I'm having trouble distinguishing them from a regular null from a common sequence.  The sub-sub-arrays need the subArrayOffset subtracted from their startOffset so the the time match is found.

I'm thinkings that in events of a sub-sub-array if the startOffset is updated with the -subArrayOffset then they might be recognized for removal and possibly work for deeper nesting also.  But I can't seem to get a reference to the event.

NOTE: Currently using a global var subArrayOffset defined outside of Tone.Part.prototype.remove() 

I've updated the Demo page so show the current behavior and I'm checking that I'm not breaking remove() for Tone.Part.

http://www.guitarland.com/ExperimentWithSeq.html

getting closer.
